File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/domain/module/TenantModule.java
package br.com.deltaglobalbank.identity.domain.module;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public class TenantModule {

    private final UUID id;
    private final UUID tenantId;
    private final UUID moduleId;
    private final Instant enabledAt;
    private final Instant deletedAt;

    private boolean enabled;
    private Instant updatedAt;

    public TenantModule(
        UUID id,
        UUID tenantId,
        UUID moduleId,
        boolean enabled,
        Instant enabledAt,
        Instant updatedAt,
        Instant deletedAt
    ) {
        this.id = id;
        this.tenantId = tenantId;
        this.moduleId = moduleId;
        this.enabled = enabled;
        this.enabledAt = enabledAt;
        this.updatedAt = updatedAt;
        this.deletedAt = deletedAt;
    }

    public static TenantModule create(UUID id, UUID tenantId, UUID moduleId) {
        Instant now = Instant.now();
        return new TenantModule(
            id,
            tenantId,
            moduleId,
            true,
            now,
            now,
            null
        );
    }

    public UUID getId() {
        return id;
    }

    public UUID getTenantId() {
        return tenantId;
    }

    public UUID getModuleId() {
        return moduleId;
    }

    public Instant getEnabledAt() {
        return enabledAt;
    }

    public Instant getDeletedAt() {
        return deletedAt;
    }

    public void enable() {
        enabled = true;
        updatedAt = Instant.now();
    }

    public void disable() {
        enabled = false;
        updatedAt = Instant.now();
    }

    public boolean isEnabled() {
        return enabled;
    }

    public TenantModuleSnapshot snapshot() {
        return new TenantModuleSnapshot(
            id,
            tenantId,
            moduleId,
            enabled,
            enabledAt,
            updatedAt,
            deletedAt
        );
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof TenantModule tenantModule && id.equals(tenantModule.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "TenantModule(id=" + id + ", tenantId=" + tenantId + ", moduleId=" + moduleId + ", enabled=" + enabled + ")";
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/domain/module/TenantModule.kt
package br.com.deltaglobalbank.identity.domain.module

import java.time.Instant
import java.util.UUID

class TenantModule(
    val id: UUID,
    val tenantId: UUID,
    val moduleId: UUID,
    enabled: Boolean,
    val enabledAt: Instant,
    updatedAt: Instant,
    val deletedAt: Instant?
) {

    companion object {
        fun create(id: UUID, tenantId: UUID, moduleId: UUID): TenantModule {
            val now = Instant.now()
            return TenantModule(
                id = id,
                tenantId = tenantId,
                moduleId = moduleId,
                enabled = true,
                enabledAt = now,
                updatedAt = now,
                deletedAt = null
            )
        }
    }

    private var _enabled: Boolean = enabled
    private var _updatedAt: Instant = updatedAt

    fun enable() {
        _enabled = true
        _updatedAt = Instant.now()
    }

    fun disable() {
        _enabled = false
        _updatedAt = Instant.now()
    }

    fun isEnabled(): Boolean = _enabled

    fun snapshot(): TenantModuleSnapshot = TenantModuleSnapshot(
        id = id,
        tenantId = tenantId,
        moduleId = moduleId,
        enabled = _enabled,
        enabledAt = enabledAt,
        updatedAt = _updatedAt,
        deletedAt = deletedAt
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TenantModule) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String =
        "TenantModule(id=$id, tenantId=$tenantId, moduleId=$moduleId, enabled=$_enabled)"
}

data class TenantModuleSnapshot(
    val id: UUID,
    val tenantId: UUID,
    val moduleId: UUID,
    val enabled: Boolean,
    val enabledAt: Instant,
    val updatedAt: Instant,
    val deletedAt: Instant?
)
